### PR TITLE
refactor: generalize watermark fetching as an interface of ISB service. Fixes #252

### LINKED ISSUE #252: Do not hard code to use JetStream in pipeline_watermark_query.go
1. Do not hard code to use JetStream
2. Single jetstream connection, and need to handle `close`

_Originally posted by @whynowy in https://github.com/numaproj/numaflow/pull/120#discussion_r927271081

### PR DESCRIPTION
**Pull request by KeranYang**
### Description
Fixes #252 This change is to NOT hardcode JetStream as ISB service when fetching watermarks. When the ISB service is Redis, we use noop KV watchers(for both heartbeat and offset timeline) for fetching watermarks.

### Testing
**Before the change**, if we use Redis as ISB service, when creating a simple-pipeline, the `simple-pipeline-daemon-xxx` pod crashes with following error:

```
Error: failed to create grpc server: failed to create watermark fetcher  failed to get nats connection, environment variable "NUMAFLOW_ISBSVC_JETSTREAM_URL" not found
Usage:
  numaflow daemon-server [flags]

Flags:
  -h, --help                 help for daemon-server
      --isbsvc-type string   ISB Service type, e.g. jetstream (default "jetstream")

panic: failed to create grpc server: failed to create watermark fetcher  failed to get nats connection, environment variable "NUMAFLOW_ISBSVC_JETSTREAM_URL" not found

goroutine 1 [running]:
github.com/numaproj/numaflow/cmd/commands.Execute()
	/Users/kyang5/Desktop/development/numaflow/numaflow/cmd/commands/root.go:17 +0x45
main.main()
	/Users/kyang5/Desktop/development/numaflow/numaflow/cmd/main.go:8 +0x17
```

And numaflow UI shows empty page for simple-pipeline.

**After the change**, when using Redis as ISB, we can start the `simple-pipeline-daemon-xxx` pod successfully. On the UI, we show default watermark -1 across ALL vertices.
redis

Signed-off-by: Keran Yang <yangkr920208@gmail.com>

Explain what this PR does.

### CODE REVIEW

**Comment by KeranYang** [comment]
On issue #252 , there is also the following request 
> 2. Single jetstream connection, and need to handle close.

@whynowy Hey Derek, regarding the request above, is it already resolved by https://github.com/numaproj/numaflow/pull/238/files?

File: pkg/isbsvc/jetstream_service.go [hunk 1]
```diff
@@ -21,6 +25,21 @@ type jetStreamSvc struct {
 	js       *jsclient.JetStreamContext
 }
 
+func (jss *jetStreamSvc) CreateWatermarkFetcher(ctx context.Context, pipelineName string, bufferName string) (fetch.Fetcher, error) {
```
Comment (whynowy) [inline pkg/isbsvc/jetstream_service.go:28]: `pipelineName` is already a property of the struct, not needed in the function signature.

File: pkg/isbsvc/jetstream_service.go [hunk 2]
```diff
@@ -21,6 +25,21 @@ type jetStreamSvc struct {
...
+	hbBucketName := JetStreamProcessorBucket(pipelineName, bufferName)
+	hbWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, hbBucketName, jsclient.NewInClusterJetStreamClient())
```
Comment (whynowy) [inline pkg/isbsvc/jetstream_service.go:30]: Use the jetstream client in the struct.

File: pkg/isbsvc/jetstream_service.go [hunk 3]
```diff
@@ -21,6 +25,21 @@ type jetStreamSvc struct {
...
+	if err != nil {
+		return nil, err
+	}
+	otBucketName := JetStreamOTBucket(pipelineName, bufferName)
+	otWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, otBucketName, jsclient.NewInClusterJetStreamClient())
```
Comment (whynowy) [inline pkg/isbsvc/jetstream_service.go:35]: same here.

File: pkg/isbsvc/jetstream_service.go (hunk 1 above)
Comment (KeranYang) [reply pkg/isbsvc/jetstream_service.go]: Thanks! Addressed.

**Review by vigith — APPROVED** [review]

File: pkg/isbsvc/jetstream_service.go [hunk 4]
```diff
@@ -25,14 +25,14 @@ type jetStreamSvc struct {
 	js       *jsclient.JetStreamContext
 }
 
-func (jss *jetStreamSvc) CreateWatermarkFetcher(ctx context.Context, pipelineName string, bufferName string) (fetch.Fetcher, error) {
-	hbBucketName := JetStreamProcessorBucket(pipelineName, bufferName)
-	hbWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, hbBucketName, jsclient.NewInClusterJetStreamClient())
+func (jss *jetStreamSvc) CreateWatermarkFetcher(ctx context.Context, bufferName string) (fetch.Fetcher, error) {
+	hbBucketName := JetStreamProcessorBucket(jss.pipelineName, bufferName)
+	hbWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, jss.pipelineName, hbBucketName, jsclient.NewInClusterJetStreamClient())
```
[the remark this answers is not in the document]
Comment (KeranYang) [reply pkg/isbsvc/jetstream_service.go:30]: Good catch. There are two types of jsClient - `defaultJetStreamClient` and `inClusterJetStreamClient`. Searching across the code base I see `defaultJetStreamClient` is only used in unit test files so I think it's safe to get jsClient from jss.jsClient. Will address.

File: pkg/isbsvc/jetstream_service.go [hunk 5]
```diff
@@ -21,6 +25,21 @@ type jetStreamSvc struct {
 	js       *jsclient.JetStreamContext
 }
 
+func (jss *jetStreamSvc) CreateWatermarkFetcher(ctx context.Context, bufferName string) (fetch.Fetcher, error) {
```
Comment (whynowy) [inline pkg/isbsvc/jetstream_service.go:28]: Nit: move this function after `NewISBJetStreamSvc(`, let it sit together with other interface functions.

Comment (KeranYang) [reply pkg/isbsvc/jetstream_service.go]: Sure, addressed.

**Review by vigith — APPROVED** [review]

**Pull request merged**